finality-aleph/src/testing/data_io.rs

```diff
@@ -0,0 +1,385 @@
+    assert_eq!(message.included_blocks(), blocks);
+
+    exit_data_store_tx.send(()).unwrap();
+    data_store_handle.await.unwrap();
+}
+
+#[tokio::test]
+async fn sends_messages_after_import() {
+    let (
+        task_handle,
+        mut client,
+        DataStoreChannels {
+            mut network,
+            block_requests_rx: _,
+            network_tx,
+            exit_data_store_tx,
+        },
+    ) = prepare_data_store();
+    let data_store_handle = tokio::spawn(task_handle);
+
+    // Build block
+    let block = client
+        .new_block(Default::default())
+        .unwrap()
+        .build()
+        .unwrap()
+        .block;
+
+    let data = AlephData::new(block.header.hash(), block.header.number);
+
+    // Send created block
+    network_tx.unbounded_send(vec![data]).unwrap();
+
+    // Import created block
+    client
+        .import(BlockOrigin::Own, block.clone())
+        .await
+        .unwrap();
+
+    // Block should be sent to network
+    let message = network.next().await.unwrap();
+    assert_eq!(message.included_blocks(), vec![data]);
+
+    exit_data_store_tx.send(()).unwrap();
+    data_store_handle.await.unwrap();
+}
+
+#[tokio::test]
+async fn sends_messages_with_number_lower_than_finalized() {
+    let (
+        task_handle,
+        mut client,
+        DataStoreChannels {
+            mut network,
+            block_requests_rx: _,
+            network_tx,
+            exit_data_store_tx,
+        },
+    ) = prepare_data_store();
+    let data_store_handle = tokio::spawn(task_handle);
+
+    // Finalize 4 blocks
+    import_blocks(&mut client, 4, true).await;
+
+    // Build 4 new blocks different from finalized
+    let mut blocks = Vec::new();
+    for i in 0u64..4 {
+        let mut digest = Digest::default();
+        digest.push(sp_runtime::generic::DigestItem::Other(
+            i.to_le_bytes().to_vec(),
+        ));
+        let block = client
+            .new_block_at(&BlockId::Number(i), digest, false)
+            .unwrap()
+            .build()
+            .unwrap()
+            .block;
+        blocks.push(AlephData::new(block.header.hash(), block.header.number));
+    }
+
+    // Send the created blocks
+    network_tx.unbounded_send(blocks.clone()).unwrap();
+
+    // Blocks should be sent to network
+    let message = network.next().await.unwrap();
+    assert_eq!(message.included_blocks(), blocks);
+
+    exit_data_store_tx.send(()).unwrap();
+    data_store_handle.await.unwrap();
+}
+
+#[tokio::test]
+async fn does_not_send_messages_without_import() {
+    let (
+        task_handle,
+        mut client,
+        DataStoreChannels {
+            mut network,
+            block_requests_rx: _,
+            network_tx,
+            exit_data_store_tx,
+        },
+    ) = prepare_data_store();
+    let data_store_handle = tokio::spawn(task_handle);
+
+    // Import 4 blocks
+    let blocks = import_blocks(&mut client, 4, false).await;
```

Comment:
I can see those comments were now present in the previous file version, hence wanted to ask if they are required? E.g. this one reader can see clearly that we import 4 blocks here (ie name of function `import blocks` fits behavior here). Not really a big deal, but wanted to clarify whether there are any testing code convention I'm not aware of.

This is a general rule I learned in the Robert C Martin videos in the past - if I need the comment I usually extract method so method name says what the code does etc., instead of putting comment.

Reply:
This comment applies to other places as well, but leaving it into Author's consideration :)

Reply:
The comment about importing is definitely unnecessary. I added all those comments, do describe the scenarios tested as code might be a little bit convolutet in some places. Tbh the names of tests are descriptive enough, so I can just remove them.

Reply:
Removed all comments as now everything is less spaghettier code.